libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotter.kt

```diff
@@ -5,111 +5,167 @@
+                // An anonymous or synthetic class also can't be referenced from other source files, so its snapshot is also empty.
+                EmptyJavaClassSnapshot
             } else {
-                JarContentsReader(classpathEntry)
+                null
             }
         }
+
+        // Snapshot the remaining classes in one invocation
+        val remainingClassesIndices: List<Int> = classFiles.indices.filter { specialCaseSnapshots[it] == null }
+        val remainingClassIds: List<ClassId> = remainingClassesIndices.map { classIds[it] }
+        val remainingClassesContents: List<ByteArray> = remainingClassesIndices.map { classes[it].contents }
+
+        val snapshots: List<JavaClassSnapshot> = JavaClassDescriptorCreator.create(remainingClassIds, remainingClassesContents).map {
+            RegularJavaClassSnapshot(it.toSerializedJavaClass())
+        }
+        val remainingSnapshots: Map<Int, JavaClassSnapshot> /* maps a class index to its snapshot */ =
+            remainingClassesIndices.zip(snapshots).toMap()
+
+        // Return a snapshot for each class
+        return classFiles.indices.map { specialCaseSnapshots[it] ?: remainingSnapshots[it]!! }
     }
+}
+
+/** Utility to read the contents of a directory or jar. */
+private object DirectoryOrJarContentsReader {
 
     /**
-     * Returns a map from (Unix-like) relative paths of classes to their contents. The paths are relative to the containing classpath entry
-     * (directory or jar).
+     * Returns a map from Unix-style relative paths of entries to their contents. The paths are relative to the container (directory or
+     * jar).
      *
      * The map entries need to satisfy the given filter.
      *
-     * The map entries are sorted based on their (Unix-like) relative paths (to ensure deterministic results across filesystems).
+     * The map entries are sorted based on their Unix-style relative paths (to ensure deterministic results across filesystems).
      *
      * Note: If a jar has duplicate entries, only one of them will be used (there is no guarantee which one will be used).
      */
-    abstract fun readContents(filter: ((invariantSeparatorsRelativePath: String, isDirectory: Boolean) -> Boolean)? = null):
-            LinkedHashMap<String, ByteArray>
-}
-
-/** Utility to read the contents of a directory. */
-class DirectoryContentsReader(private val directory: File) : ClasspathEntryContentsReader() {
-
-    init {
-        check(directory.isDirectory)
+    fun read(
+        directoryOrJar: File,
+        entryFilter: ((unixStyleRelativePath: String, isDirectory: Boolean) -> Boolean)? = null
+    ): LinkedHashMap<String, ByteArray> {
+        return if (directoryOrJar.isDirectory) {
+            readDirectory(directoryOrJar, entryFilter)
+        } else {
+            check(directoryOrJar.isFile && directoryOrJar.path.endsWith(".jar", ignoreCase = true))
```

Comment:
I see. However, there are many places in AGP & KGP where we handle only jars or directories, not zip files (e.g., artifact transforms are mostly/always created for artifact type `jar`, not `zip`).

I can change this code to allow `.zip` but that means it will hide the issue (that `.zip` files are not yet handled). Alternatively, we can audit the code to ensure `.zip` files are handled whenever jars are handled, but that seems like a significant effort, and I'm not aware of any bug reports on this.

So maybe it's actually good that we fail early here?

Reply:
(Btw, currently we will not see `.zip` files here because this code is called from `ClasspathEntrySnapshotTransform` which operates on directories and jars only, due to the way artifact transforms are currently set up.)

Reply:
What is the current behavior if you add  a `.zip` in the classpath?

Reply:
It will be ignored. `KotlinCompile.classpath` will not contain that zip file (I just checked), same for classpath snapshots as classpath snapshot files match classpath files.

Reply:
SGTM. As a precaution, let's add a TODO to add an integration tests which has .zip in the classpath, just to make things do not fall apart.

Reply:
Sure, I'll make a note.